File: Sweep Line Algorithm/218. The Skyline Problem/solution.py
#https://leetcode.com/problems/the-skyline-problem/

class Solution:
    def getSkyline(self, buildings: List[List[int]]) -> List[List[int]]:
        length = len(buildings)
        if length == 2:
            prev = buildings[0]
            tail = buildings[1]
            answer = []
            if prev[1] < tail[0]: #两个根本不重叠
                answer.append([prev[0], prev[2]])
                answer.append([prev[1] ,0])
                answer.append([tail[0], tail[2]])
                answer.append([tail[1], 0])
                return answer
            elif prev[1] >= tail[1]:
                if prev[2] >= tail[2]:
                    answer.append([prev[0], prev[2]])
                    answer.append([prev[1], 0])
                    return answer
                else:
                    answer.append([prev[0], prev[2]])
                    answer.append([tail[0], tail[2]])
                    if prev[1] != tail[1]:
                        answer.append([tail[1], prev[2]])
                    answer.append([prev[1], 0])
                    return answer
            else:
                answer.append([prev[0], prev[2]])
                if prev[2] > tail[2]:
                    answer.append([prev[1], tail[2]])
                elif prev[2] < tail[2]:
                    answer.append([tail[0], tail[2]])
                else:
                    pass
                answer.append([tail[1], 0])
                return answer
        else:
            list1 = self.getSkyline(buildings[0:int (length/2)])
            list2 = self.getSkyline(buildings[int (length/2+1):])
            return merge(list1, list2)
    def merge(self, prev: List[List[int]], tail: List[List[int]])-> List[List[int]]:
        list = prev + tail
        new_prev = []
        new_tail = []
        for each in prev:
            new_prev.append[each[0], each[1], 1]
        for each in tail:
            new_tail.append[each[0], each[1], -1]
        list = new_prev + new_tail
        list.sort()
        prev_height = 0
        tail_height = 0
        answer = []
        for each in list:
            if each[2] == 1: # 是前一个轮廓的点
                prev_height = each[1]
                if prev_height > tail_height:
                    answer.append([each[0], each[1]])
            if each[2] == -1: # 是后一个轮廓的点
                tail_height = each[1]
                if tail_height > prev_height:
                    answer.append([each[0], each[1]])

File: Sweep Line Algorithm/218. The Skyline Problem/solution.cpp
//https://leetcode.com/problems/the-skyline-problem/
#include <vector>
#include <iostream>
using namespace std;
class Solution {
public:
    vector<vector<int>> getSkyline(vector<vector<int>>& buildings) {
        // events,   x,   h,   id,  type (1=entering, -1=leaving)
        vector<Event> events;
        int id = 0;
        for(const auto& b : buildings){
            events.push_back(Event{id, b[0], b[2], 1}); //进入事件
            events.push_back(Event{id++, b[1], b[2], -1}); //离开事件
        }
        sort(events.begin(), events.end());

        vector<vector<int>> ans;

    }
private:
    struct Event{
        int id;
        int x;
        int h;
        int type;

        bool operator < (const Event& e) const{
            if (x == e.x)
                // 如果是进入事件x重合，那么先考虑更高的 (type = 1)
                // 如果是离开事件x重合，那么优先考虑更矮的 (type = -1)
                return type * h > e.type * e.h;
            return x < e.x;
        }
    }
};
